src/mpl/src/gpu/mpl_gpu_cuda.c

```diff
@@ -15,13 +15,39 @@ typedef struct gpu_free_hook {
     struct gpu_free_hook *next;
 } gpu_free_hook_s;
 
+static int device_count = -1;
+static int max_dev_id = -1;
+
+static cuda_gpu_attr_s gpu_attributes;
 static gpu_free_hook_s *free_hook_chain = NULL;
 
 static CUresult CUDAAPI(*sys_cuMemFree) (CUdeviceptr dptr);
 static cudaError_t CUDARTAPI(*sys_cudaFree) (void *dptr);
 
 static int gpu_mem_hook_init();
 
+int MPL_gpu_get_dev_count(int *dev_cnt)
+{
+    if (device_count == -1) {
+        cudaError_t ret = cudaGetDeviceCount(&device_count);
+        CUDA_ERR_CHECK(ret);
+        *dev_cnt = device_count;
+    } else {
+        *dev_cnt = device_count;
+    }
+
+  fn_exit:
+    return MPL_SUCCESS;
+  fn_fail:
+    return MPL_ERR_GPU_INTERNAL;
+}
+
+int MPL_gpu_get_max_dev_id(int *dev_id)
+{
+    *dev_id = max_dev_id;
```

Comment:
Why do you leave the code getting `max_dev_id` in `MPL_gpu_init`?

Reply:
I can move it to `MPL_gpu_get_max_dev_id`.
In this case, there would be no stuff in `MPL_gpu_init` :)
what is `MPL_gpu_init` for?

Reply:
I think I cannot move `cudaGetDeviceCount` into `MPL_gpu_get_dev_count`, because `max_dev_id` depends on device count. I would keep my original scheme. How do you think?

Reply:
ZE needs init. And your cuda mem hooks needs init. 

It is not a bad thing to have empty init or not needing to init. Actually, how about add a `static bool gpu_initialized;`, then in the functions that depends on init, for example, in the ze `MPL_gpu_get_dev_count`, it can go like:
```
if (!gpu_initialized) {
    MPL_gpu_init();
}
return dev_count;
```

And in case of cuda, e.g. when it doesn't need init at all, you can simply skip the whole thing.

That way, you can remove the gpu init from MPIR_Init_thread, and keep `MPL_gpu_init()` internally static. Right?

Reply:
> I think I cannot move `cudaGetDeviceCount` into `MPL_gpu_get_dev_count`, because `max_dev_id` depends on device count. I would keep my original scheme. How do you think?

Of course. Will `int MPL_gpu_get_dev_count(int *dev_count, int *max_id)` work?

Reply:
sounds good